### PR TITLE
Emit answer rework

### PR DESCRIPTION
**Pull request by shenjackyuanjie**
a rework of #310 
WARNING: this PR is not fully tested

please be free to add any commit to this PR

### CODE REVIEW

File: socketio/src/asynchronous/socket.rs [hunk 1]
```diff
@@ -84,11 +84,61 @@ impl Socket {
...
+                Ok(Packet::new(
+                    package_type,
+                    nsp.to_owned(),
+                    Some(payload),
+                    id,
+                    0,
+                    None,
+                ))
+            },
+            Payload::Text(data) => todo!("not implemented yet"),
```
Comment (shenjackyuanjie) [inline socketio/src/asynchronous/socket.rs:138]: first todo

File: socketio/src/socket.rs [hunk 2]
```diff
@@ -73,11 +73,68 @@ impl Socket {
...
+                Ok(Packet::new(
+                    package_type,
+                    nsp.to_owned(),
+                    Some(payload),
+                    id,
+                    0,
+                    None,
+                ))
+            },
+            Payload::Text(data) => todo!("Text: {:?}", data),
```
Comment (shenjackyuanjie) [inline socketio/src/socket.rs:134]: second todo

**Review by shenjackyuanjie — COMMENTED** [review]
the two todo I don't know how to do it

**Comment by shenjackyuanjie** [comment]
there is still some conflict found with #384

**Comment by shenjackyuanjie** [comment]
@1c3t3a  hello?(

**Comment by 1c3t3a** [comment]
Hey @shenjackyuanjie, I'll try and get to it over the holidays!

**Comment by shenjackyuanjie** [comment]
> Hey @shenjackyuanjie, I'll try and get to it over the holidays!

got it! have a nice day~

**Comment by shenjackyuanjie** [comment]
and I think I should fix the test first(

**Comment by shenjackyuanjie** [comment]
I just remove the new test, which will cause issue, and can't test any new feature(

**Comment by shenjackyuanjie** [comment]
@1c3t3a  if is available, I think you can remove the workflows approve(
since I can't run test locally

**Comment by 1c3t3a** [comment]
Seems like this is still failing a bunch of doc tests.

**Comment by 1c3t3a** [comment]
Actually I am not very happy with the implementation here. As stated in the original issue, I am not sure whether this is really needed: (see #95). Also this starts to expose the Ack ID to *every* callback we ever define and reworks *all* our public callback APIs. Until now, the Ack ID is never exposed and an implementation detail that is not exposed to the outside world. If we would ever want to expose this, I would rather introduce a new new Callback type that allows for that functionality and pass that to selected public APIs.

In the current state this is changing all public APIs, which I don't think is reasonable for such a small feature.

**Comment by Totodore** [comment]
I'm not really familiar to this codebase but I have another proposition, because the `ack` id is related to the received payload. Why can't we create an `AckSender` struct that is part of the payload and that contains a `Clone` of `InnerSocket` and the ack id.
The `AckSender` has an `emit` method to "answer" with an `acknowledgement` corresponding to the ack id. 

```rust
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Payload {
    Binary(Bytes),
    String(String),
    StringAck(String, AckSender),
    BinaryAck(Bytes, AckSender),
}
```
It is basically what I did here for the [socketio server implementation](https://github.com/Totodore/socketioxide/blob/39bc2f9d8daad4ad8f7cbe411df6a24ca35f4124/socketioxide/src/handler/extract.rs#L247-L254) : 
```rust
/// An Extractor to send an ack response corresponding to the current event.
/// If the client sent a normal message without expecting an ack, the ack callback will do nothing.
#[derive(Debug)]
pub struct AckSender<A: Adapter = LocalAdapter> {
    binary: Vec<Vec<u8>>,
    socket: Arc<Socket<A>>,
    ack_id: Option<i64>,
}
...
pub fn send(self, data: impl Serialize) -> Result<(), SendError> {
    if let Some(ack_id) = self.ack_id {
        let ns = self.socket.ns();
        let data = serde_json::to_value(&data)?;
        let packet = if self.binary.is_empty() {
            Packet::ack(ns, data, ack_id)
        } else {
            Packet::bin_ack(ns, data, self.binary, ack_id)
        };
        self.socket.send(packet)
    } else {
        Ok(())
    }
}
```

Whatever the way it is implemented, it would be really helpful to have this feature :)

**Comment by shenjackyuanjie** [comment]
@1c3t3a how about I spilt this PR into two
or just remove the Ack part?
for now, master branch still can't receive multiple payload at once

**Pull request closed without merging**